### DISCUSSION TITLE
Pages Router router.push() sometimes performs unexpected full page refresh

### DISCUSSION THREAD

**Discussion by yss14** [vercel/next.js#58302] [Help]
### Summary

_Disclaimer: Since the changed behaviour of the router appears kind of random and flaky to me I was not able to create a simplistic reproduction scenario, and thus cannot report an official bug report._

I ran into this weird behaviour of `router.push()` that under certain unknown circumstances it performs a full page refresh (so the root `App` component in `_app.tsx` is mounting again) and under some not (as I would expect it)?
I further narrowed it down to [v13.4.3](https://github.com/vercel/next.js/releases/tag/v13.4.3) introducing this (breaking) change, since with v13.4.2 everything still works fine, so every `router.push()` call results in a normal (SPA) page transition.
I already had a look at the changelog and which files changed. Apparently, there are two small changes to the client router in this particular version. One [reworking client router filter handling](https://github.com/vercel/next.js/pull/49741) and another [fixing some behaviour with hash links](https://github.com/vercel/next.js/pull/50006).
https://github.com/vercel/next.js/compare/v13.4.2...v13.4.3

Correct behaviour with v13.4.2:

https://github.com/vercel/next.js/assets/4346842/9554ff96-e146-46c3-81a9-2821f6a1710e

Broken behaviour with v13.4.3:

https://github.com/vercel/next.js/assets/4346842/2f86a99b-b98f-4bd3-a643-0dea40abaf6e

My code which is performing the `router.push()` calls looks like the following.

```typescript
import { useLocale } from 'next-intl'
import { useRouter } from 'next/router'

export const useSubmitFormView = () => {
  const { push } = useRouter()
  const locale = useLocale()

  return () => {
    const nextView = /*some code evaluating value of <nextView>
    if (nextView) {
        console.log(
          'Link to',
          prefixLocaleLink(`/booking-journey/${product.id}/configuration/${nextView._key}`, locale)
        )
        push(prefixLocaleLink(`/booking-journey/${product.id}/configuration/${nextView._key}`, locale))
      } else {
        console.log('Link to', prefixLocaleLink(`/booking-journey/overview`, locale))
        push(prefixLocaleLink(`/booking-journey/overview`, locale))
      }
  }
}

const prefixLocaleLink = (link: `/${string}`, locale: string) => {
  return `/${locale}${link}`
}
```

To debug this whole scenario I'm also listening to the `router.events` in my root `App` component like this.

```typescript
export const usePageRouterEventLogs = () => {
  const router = useRouter()

  useEffect(() => {
    const onRouteChangeStart = (...args: any[]) => {
      console.debug('[ROUTER]: routeChangeStart', args)
    }
    const onRouteChangeComplete = (...args: any[]) => {
      console.debug('[ROUTER]: routeChangeComplete', args)
    }
    const onRouteChangeError = (...args: any[]) => {
      console.debug('[ROUTER]: routeChangeError', args)
    }

    router.events.on('routeChangeStart', onRouteChangeStart)
    router.events.on('routeChangeComplete', onRouteChangeComplete)
    router.events.on('routeChangeError', onRouteChangeError)

    return () => {
      router.events.off('routeChangeStart', onRouteChangeStart)
      router.events.off('routeChangeComplete', onRouteChangeComplete)
      router.events.off('routeChangeError', onRouteChangeError)
    }
  }, [router.events])
}
```

Weird thing I recognized is, that in case when a full page refresh is triggered by my `router.push()` call no router event is even logged, whereas in cases when nextjs performs a "normal" page transition (as you would expect it from a SPA) both `routeChangeStart` and `routeChangeComplete` are triggered as expected.Further, I recognized that for some transitions it works without re-rendering the whole app, whereas for other page transition triggered by the exact same logic a full page refresh is performed.

Does anyone who is more familiar with code of nextjs unterstand and know why v13.4.3 breaks this behaviour and if so, which change in particular?

**Comment by icyJoseph** [comment]
Oh wow, this one gives anxiety vibes...

Do you have any catch all routes? rewrites? redirects?

**Reply by yss14** [reply]
@icyJoseph This is my `next.config.js`

```javascript
module.exports =
  withNextIntl(
    withBundleAnalyzer({
      reactStrictMode: true,
      images: {
        domains: ['cdn.sanity.io'],
      },
      async redirects() {
        return [
          {
            // historical change to preserve customer saved deeplinks
            // https://nextjs.org/docs/pages/api-reference/next-config-js/redirects#wildcard-path-matching
            source: '/bookingJourney/:slug*',
            destination: '/booking-journey/:slug*',
            permanent: true,
          },
        ]
      },
      reactStrictMode: false,
    })
  )
```

Already tried with disabled redirects in the config, but same result.

Catch all routes not for pages, but some api routes are catch all routes under the `/api` directory. Don't know if that would infer somehow?
Rewrites only in our middleware, which I already deactivated. 
Redirects are also all commented out, but only located in api routes.

**Comment by Baalamurgan** [comment]
Is there a solution for this yet?

**Reply by yss14** [reply]
Haven't found any. We ended up persisting our state in sessionState as a workaround.

**Reply by kumar0** [reply]
Using N14, I am also having the same issue... it works well locally  yarn dev and yarn build/start.. when I hosted to fargate , it refreshes the page

**Reply by abhishek-solaris** [reply]
@kumar0 Did you found the solution to it ? I am also facing the similar issue.

**Comment by qingzhoufeihu** [comment]
The situation I encountered is that when I jump back and forth between the list page and the details page using router.push more times, the rendering speed of the page switching becomes slower and slower. After the URL changes, it takes 5 seconds to render the list page.

**Comment by pablojsx** [comment]
april 2024 same issue, it will reload the page. I don't wanna use redux or add code somewhere in other files man...

**Comment by madsgodvinjensen** [comment]
I had a dive into this as we were seeing issues with _some_ paths. Not all. It seems to have to do with this feature: https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#7-routing-between-pages-and-app

Somehow, some paths get matched up against a filter in the router code base - and that ends up in a [code branch](https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/router/router.ts#L1052) that runs `window.location.href = ...`. 

For us, turning off the filter (next.config.js -> `experimental.clientRouterFilter: false`) did the trick.

We _did_ also have a `./src/app` folder with one thing using App Router. Taking that out of the app and removing the app folder also fixed the problem...

**Reply by ManuelAngel99** [reply]
Thank you so much, we had this exact issue and this fixed it!

For SEO purposes, if anyone is experiencing random nexjs refreshes when clicking a link to a dynamic route, this solves the problem.

**Reply by kaihenthorniwane** [reply]
Yep, pages router was randomly full page refreshing some routes with no clear pattern, but disabling this fixed it.

**Comment by Wadan3** [comment]
Hi! This unexpected full page refresh issue in `v13.4.3` was indeed caused by the rewrite and filtering changes made to the client-side router (specifically how it handles internal route filters and trailing slashes/hashes). When the router fails to client-side match the target URL structure internally, it panics and falls back to a hard window location refresh.

Here are the tested ways to fix or work around this behavior:

### 1. Upgrade Next.js (Recommended)
This was a known regression introduced in `v13.4.3`. It was subsequently patched in later minor versions (v13.4.4+ and v14/v15). Upgrading your project dependency is the cleanest solution:
```bash
npm install next@latest react@latest react-dom@latest
